packages/components/src/components/PasswordCreationField/lib/usePolicyValidationResult.ts

```diff
@@ -0,0 +1,59 @@
+import { startTransition, useEffect } from "react";
+import type { PolicyValidationResult } from "@/integrations/@mittwald/password-tools-js";
+import { isPromise } from "remeda";
+import type { Policy } from "@/integrations/@mittwald/password-tools-js";
+import type { ResolvedPolicyValidationResult } from "@/components/PasswordCreationField/PasswordCreationField";
+
+export const usePolicyValidationResult = (
+  validationPolicy: Policy,
+  password: string,
+  onValidationStart?: () => void,
+  onValidationResult?: (data: {
+    password: string;
+    isValid: boolean;
+    results: ResolvedPolicyValidationResult;
+  }) => void,
+) => {
+  useEffect(() => {
+    onValidationStart?.();
+    const validationResult = validationPolicy.validate(password);
+
+    const setValidationResult = (
+      password: string,
+      policyValidationResult: PolicyValidationResult,
+    ) => {
+      const isValid = Boolean(policyValidationResult.isValid);
+      onValidationResult?.({
+        password,
+        isValid,
+        results: policyValidationResult as ResolvedPolicyValidationResult,
+      });
+    };
+
+    startTransition(async () => {
+      if (!isPromise(validationResult.isValid)) {
+        return setValidationResult(password, validationResult);
+      }
+
+      void Promise.all([
+        Promise.resolve(password),
+        Promise.resolve(validationResult),
+        ...validationResult.ruleResults,
+      ]).then(
+        ([
+          resolvedValue,
+          resolvedValidationResult,
+          ...resolvedValidationRuleResults
+        ]) => {
+          startTransition(() => {
+            setValidationResult(resolvedValue, {
+              complexity: resolvedValidationResult.complexity,
+              ruleResults: resolvedValidationRuleResults,
+              isValid: !resolvedValidationRuleResults.some((r) => !r.isValid),
+            });
+          });
+        },
+      );
+    });
+  }, [password]);
```

Comment:
Should we maybe also watch for policy or callback changes?
```suggestion
  }, [password, validationPolicy, onValidationStart, onValidationResult]);
```